src/NHibernate/Impl/SessionFactoryImpl.cs

```diff
@@ -1270,5 +1267,144 @@ public string Uuid
 		}
 
 		#endregion
+
+		// NH specific: implementing return type covariance with interface is a mess in .Net.
+		internal class SessionBuilderImpl : SessionBuilderImpl<ISessionBuilder>, ISessionBuilder
+		{
+			public SessionBuilderImpl(SessionFactoryImpl sessionFactory) : base(sessionFactory)
+			{
+				SetSelf(this);
+			}
+		}
+
+		internal class SessionBuilderImpl<T> : ISessionBuilder<T>, ISessionCreationOptions where T : ISessionBuilder<T>
+		{
+			// NH specific: implementing return type covariance with interface is a mess in .Net.
+			private T _this;
+			private static readonly IInternalLogger _log = LoggerProvider.LoggerFor(typeof(SessionBuilderImpl<T>));
+
+			private readonly SessionFactoryImpl _sessionFactory;
+			private IInterceptor _interceptor;
+			private DbConnection _connection;
+			// Todo: port PhysicalConnectionHandlingMode
+			private ConnectionReleaseMode _connectionReleaseMode;
+			private FlushMode _flushMode;
+			private bool _autoClose;
+
+			public SessionBuilderImpl(SessionFactoryImpl sessionFactory)
+			{
+				_sessionFactory = sessionFactory;
+
+				// set up default builder values...
+				_connectionReleaseMode = sessionFactory.Settings.ConnectionReleaseMode;
+				_autoClose = sessionFactory.Settings.IsAutoCloseSessionEnabled;
+				// NH different implementation: not using Settings.IsFlushBeforeCompletionEnabled
+				_flushMode = sessionFactory.Settings.DefaultFlushMode;
+			}
+
+			protected void SetSelf(T self)
+			{
+				_this = self;
+			}
+
+			// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+			// SessionCreationOptions
+			// NH note: it is tempting to convert them to properties, but then their names conflict
+			// with SessionBuilder interface.
+
+			public virtual FlushMode GetInitialSessionFlushMode() => _flushMode;
+
+			public virtual bool ShouldAutoClose() => _autoClose;
+
+			public DbConnection GetConnection() => _connection;
+
+			// NH different implementation: Hibernate here ignore EmptyInterceptor.Instance too, resulting
+			// in the "NoInterceptor" being unable to override a session factory interceptor.
+			public virtual IInterceptor GetInterceptor() => _interceptor ?? _sessionFactory.Interceptor;
+
+			public virtual ConnectionReleaseMode GetConnectionReleaseMode() => _connectionReleaseMode;
+
+			// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+			// SessionBuilder
+
+			public virtual ISession OpenSession()
+			{
+				_log.Debug("Opening Hibernate Session.");
+				var session = new SessionImpl(_sessionFactory, this);
+				if (_interceptor != null)
+				{
+					// NH specific feature
+					_interceptor.SetSession(session);
+				}
+
+				return session;
+			}
+
+			public virtual T Interceptor(IInterceptor interceptor)
+			{
+				// NH different implementation: Hibernate accepts null.
+				_interceptor = interceptor ?? throw new ArgumentNullException(nameof(interceptor));
```

Comment:
Makes happy some tests. Not sure we should keep this. Allowing setting null would allow for reverting a previously set session interceptor and fall-backing again on factory interceptor. Now, since builder are cheap to build, that does not look important.

Note that on Hibernate side, the `NoInterceptor` causes fallback to factory interceptor, instead of configuring a "no-op" interceptor. That does not sound good to me, the naming does not match well.